runtime/include/QuantumDevice.hpp

```diff
@@ -21,6 +21,12 @@
 #include "DataView.hpp"
 #include "Types.h"
 
+#define GENERATE_DEVICE_ENTRYPOINT(DEVICE_NAME, NAMESPACE)                                         \
+    extern "C" Catalyst::Runtime::QuantumDevice *DEVICE_NAME##Factory(const char *kwargs)          \
+    {                                                                                              \
+        return new NAMESPACE::DEVICE_NAME(std::string(kwargs));                                    \
```

Comment:
This is a handy macro, but it does slightly change the contract we have for device plugins.

The existing contract:
- the symbol of the device entry point is determined by what is returned from `get_c_interface`
- the contract is mentioned in the docs, the implementer is responsible to manually adhere to it
- the C++ device class can have any name

The new situation:
- the existing contract is still in effect (we search for the device entry point with the symbol provided by `get_c_interface`)
- but now we force the entry point symbol to correspond to the C++ device class name
- the user now has to ensure the C++ class name is the same as the device identifier returned by `get_c_interface`

This is not in itself a problem, but we have to be very clear about it in the docs, which currently only mentions the old contract.

Reply:
If this is the case, don't forget to update the docs to make this very clear!

Reply:
Good point @dime10! I think we can fix this issue with updating the signature of the template macro as follows:

``` C++
// A helper template macro to generate the <IDENTIFIER>Factory method by
// calling <CONSTRUCTOR>(kwargs).
#define GENERATE_DEVICE_FACTORY(IDENTIFIER, CONSTRUCTOR)                                           \
    extern "C" Catalyst::Runtime::QuantumDevice *IDENTIFIER##Factory(const char *kwargs)           \
    {                                                                                              \
        return new CONSTRUCTOR(std::string(kwargs));                                               \
    }
```

Where it gets both the device identifier (returned by `get_c_interface`) and the constructor of the device class. We no longer force the entry point symbol to correspond to the device class name (or constructor).  [ec8f333](https://github.com/PennyLaneAI/catalyst/pull/400/commits/ec8f3334f8f4a53cd26c71ec4873bbe64dbd9347)

Reply:
Sounds good 👍